[flang][hlfir] Pass vector subscripted elemental call arg by address

jeanPerier opened the pull request:
I missed that vector subscripted arguments must still be passed by
address in an elemental call where the dummy argument does not have
the VALUE attribute.

Update PreparedActualArgument to hold an hlfir::Entity or an
hlfir::ElementalOp and to inline the elementalOp body in `getActual`.

jeanPerier committed:
[flang][NFC] Remove getOriginalActual from PreparedActualArgument

Change in order to support vector subscripted entity passed by address
in elemental calls.

PreparedActualArguments will hold an hlfir::Entity or an
hlfir::ElementalAddrOp.

jeanPerier committed:
[flang][hlfir] Pass vector subscripted elemental call arg by address

I missed that vector subscripted arguments must still be passed by
address in an elemental call where the dummy argument does not have
the VALUE attribute.

Update PreparedActualArgument to hold an hlfir::Entity or an
hlfir::ElementalOp and to inline the elementalOp body in `getActual`.

vzakhari reviewed (approved):
Looks great! Thank you!

jeanPerier committed:
add back if removed by mistake

jeanPerier committed:
test IsVariable before HasVectorSubscript

I mistakenly thought that HasVectorSubscript would only return true for
variables, but it return true for expressions containing vector
subscripted designator in general. Fix and add a test.
